### ISSUE TITLE
Conflict between PHPStorm and phpcs PSR-12 rules

### ISSUE THREAD

**Issue by Synchro** [squizlabs/PHP_CodeSniffer#2947]
I have a persistent problem with code that is all meant to be simple PSR-12; PHPStorm and phpcs seem to have different opinions about this standard - correcting in one results in errors in the other, so there's a cycle of reformatting going on all the time. One common situation is when multi-line function calls appear in a multi-line `if` statement, for example:

```php
if (someFunction(
    1,
    'abc'
)
) {
    echo 1;
}
```
Raises no errors in phpcs, but phpcs *inside PHPStorm* marks this as:

> phpcs: PSR12.ControlStructures.ControlStructureSpacing.LineIndent: Each line in a multi-line control structure must be indented at least once; expected at least 4 spaces, but found 0

Similarly (and this is the layout I prefer, FWIW):

```php
if (someFunction(
    1,
    'abc'
)) {
    echo 1;
}
```

reports no problems in phpcs or phpcbf, but PHPStorm flags it as:

> phpcs: PSR12.ControlStructures.ControlStructureSpacing.CloseParenthesisLine: The closing parenthesis of a multi-line control structure must be on the line after the last expression

PHPStorm and phpcs are both happy with this structure:

```php
if (
    someFunction(
        1,
        'abc'
    )
) {
    echo 1;
}
```

but PHPStorm's PSR-12 reformatter reformats that to:

```php
if (
someFunction(
    1,
    'abc'
)
) {
    echo 1;
}
```

**and** then flags it with this problem:

> phpcs: PSR12.ControlStructures.ControlStructureSpacing.LineIndent: Each line in a multi-line control structure must be indented at least once; expected at least 4 spaces, but found 0

I realise this may be down to PHPStorm's config too - and [I reported it to them as such](https://youtrack.jetbrains.com/issue/WI-51892), but their response isn't very helpful (turn off formatting!) and I'm sure I can't be alone in using phpcs inside PHPStorm. To be clear, I'm using the same `phpcs.xml` config file both inside and outside PHPStorm.

Having them fight over the formatting is driving me nuts! Can you recommend a configuration for PSR-12 that works for both phpcs and PHPStorm?

**Comment by dingo-d** [comment]
I know I had the issue in PHPStorm with it not recognizing my `phpcs.xml.dist` and installed standards, maybe this could help you https://youtrack.jetbrains.com/issue/WI-45521

**Comment by Synchro** [comment]
I've done that — it's what I meant when I said:

> I'm using the same phpcs.xml config file both inside and outside PHPStorm

**Comment by gsherwood** [comment]
The latest version of PHPCS reports these error for your first sample code (on the command line):
```
$ phpcs temp.php --standard=PSR12

FILE: temp.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 1 | ERROR | [x] Header blocks must be separated by a single blank
   |       |     line
 2 | ERROR | [x] The first expression of a multi-line control
   |       |     structure must be on the line after the opening
   |       |     parenthesis
 5 | ERROR | [x] Each line in a multi-line control structure must be
   |       |     indented at least once; expected at least 4 spaces,
   |       |     but found 0
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```
PHPCBF fixes the code like this:
```php
<?php

if (
    someFunction(
        1,
        'abc'
    )
) {
    echo 1;
}
```
Which then produces no PSR12 errors.

Are you sure you are using the PSR12 standard when testing on the command line? What version of PHPCS are you running?

**Comment by Synchro** [comment]
I'm running phpcs 3.5.5. This is my config, which has some exemptions from PSR-12, but none are to do with indents:

```xml
    <rule ref="PSR12">
        <exclude name="PSR1.Methods.CamelCapsMethodName"/>
        <exclude name="PSR1.Classes.ClassDeclaration"/>
        <exclude name="PSR2.Methods.MethodDeclaration.Underscore"/>
        <exclude name="PSR2.Classes.PropertyDeclaration.Underscore"/>
        <exclude name="PSR12.Traits.UseDeclaration"/>
        <exclude name="Squiz.Classes.ValidClassName"/>
        <exclude name="Squiz.Classes.ValidClassName.NotCamelCaps"/>
    </rule>
```

I did at one point have these as well, in an attempt to work around PHPStorm's formatting, but I've now removed them, and this may explain my first two examples, however, they have no effect on PHPStorm's behaviour:

```xml
    <exclude name="PSR2.Methods.FunctionCallSignature.Indent"/>
    <exclude name="PSR12.ControlStructures.ControlStructureSpacing.FirstExpressionLine"/>
```

I think this is a bug in PHPStorm, but given how popular this combo is, I think it's worth having mention of it here too.

**Comment by Synchro** [comment]
Looks like this has been [confirmed](https://youtrack.jetbrains.com/issue/WI-50929) by someone else in PHPStorm, so I'll close this and hope that it gets fixed over there.
